### PR TITLE
fix by checking for increment

### PR DESCRIPTION
**Pull request by jasonkhadka**
## Description

Fixes [#427
](https://github.com/lf1-io/padl/issues/427)

Checks for increasing the `scopedname.n` is implemented.

### CODE REVIEW

**Comment by jasonkhadka** [comment]
Check the continued discussion on the issue: https://github.com/lf1-io/padl/issues/427

**Comment by wuhu** [comment]
Thanks for this!

Could you explain why your changes fix the problem?

Also I found that it doesn't seem to work in combination with recursive function transforms, e.g.:

```python
from padl import *
@transform
def f(x):
    return f(x-1)


f = f.pd_to('cuda')
print(f._pd_dumps())
```

raises an exception.

**Comment by jasonkhadka** [comment]
very strange!!
Cause this works:
```
from padl import *
@transform
def f(x):
    return 1 + f(x-1)


#f = f.pd_to('cuda')
print(f._pd_dumps())
```

I will have a look tomorrow, and also write the explanation.

> ```python
> from padl import *
> @transform
> def f(x):
>     return f(x-1)
> 
> 
> f = f.pd_to('cuda')
> print(f._pd_dumps())
> ```

File: padl/dumptools/var2mod.py [hunk 1]
```diff
@@ -596,8 +597,30 @@ def find_globals(node: ast.AST, filter_builtins: bool = True) -> Set[Tuple[str,
     return globals_
 
 
+def _check_and_make_increment(var: ScopedName, scope: Scope, scoped_name: ScopedName):
```
Comment (sjrl) [inline padl/dumptools/var2mod.py:600]: I have a question. Do we need both `scope` and `scoped_name` here? I thought `scoped_name` also contains the `scope` of the target.

Comment (sjrl) [reply padl/dumptools/var2mod.py:600]: Ahh I see further down how these can differ. Could you add to the `scope` doc string why this could differ from `scoped_name`?

File: padl/transforms.py [hunk 2]
```diff
@@ -526,17 +526,15 @@ def f(x):
 
             # see if the object itself knows how to generate its codegraph
             try:
-                if next_name.scope.is_global():
-                    next_obj = globals_dict[next_name.name]
-                else:
-                    next_obj = all_vars_dict[next_name.name]
+                next_obj = all_vars_dict[next_name.name]
```
Comment (sjrl) [inline padl/transforms.py:529]: @wuhu This is a change @jasonkhadka and I discussed since it looks like `all_vars_dict` already contains the `globals_dict`. Was there any other reason to have this `if` statement?

File: padl/dumptools/var2mod.py [hunk 3]
```diff
@@ -596,8 +597,30 @@ def find_globals(node: ast.AST, filter_builtins: bool = True) -> Set[Tuple[str,
...
+        new_scoped_name = ScopedName(var.name, scope, var.n + scoped_name.n + 1)
+        find_in_scope(new_scoped_name)
+    except NameNotFound:
+        new_scoped_name = ScopedName(var.name, scope, var.n + scoped_name.n)
+    return new_scoped_name
+
+
 def increment_same_name_var(variables: List[ScopedName], scoped_name: ScopedName):
-    """Go through *variables* and increment the the counter for those with the same name as
+    """Go through *variables* and increment the counter for those with the same name as
```
Comment (sjrl) [inline padl/dumptools/var2mod.py:623]: Could you add to the end of the docstring `... with the same name as *scoped_name* by incrementing *scoped_name.n*` to make it more clear what we are doing?

File: padl/transforms.py (hunk 2 above)
Comment (sjrl) [reply padl/transforms.py:529]: Okay thanks for the info! @jasonkhadka Could we add the example provided by @wuhu to the unit tests to test this scenario? Also I think it would be helpful to add a comment in the code explaining this situation or at least that `all_vars_dict` relies on `nonlocals_dict` overwriting the keys in `globals_dict`.

File: padl/dumptools/var2mod.py (hunk 3 above)
Comment (sjrl) [reply padl/dumptools/var2mod.py:623]: Never mind this doesn't make sense.

File: padl/dumptools/var2mod.py [hunk 4]
```diff
@@ -596,8 +597,30 @@ def find_globals(node: ast.AST, filter_builtins: bool = True) -> Set[Tuple[str,
     return globals_
 
 
+def _check_and_make_increment(var: ScopedName, scope: Scope, scoped_name: ScopedName):
+    """Check and make increment if the ScopedName after increment exists.
```
Comment (sjrl) [inline padl/dumptools/var2mod.py:601]: I would reword this to `Increment ScopedName and check if it exists.`

File: padl/dumptools/var2mod.py [hunk 5]
```diff
@@ -182,6 +189,7 @@ def visit_Attribute(self, node):
 
         scope = getattr(node, '_scope', None)
         name = ScopedName('.'.join(path), scope, 0)
+        # name = ScopedName('.'.join(path), scope, 0)
```
Comment (sjrl) [inline padl/dumptools/var2mod.py:192]: I think you can remove this extra comment

File: padl/dumptools/var2mod.py [hunk 6]
```diff
@@ -611,11 +619,11 @@ def _check_and_make_increment(var: ScopedName, scope: Scope, scoped_name: Scoped
     :return:
         ScopedName after increment
     """
+    new_scoped_name = ScopedName(scoped_name.name, scope, var.n + scoped_name.n + 1)
     try:
-        new_scoped_name = ScopedName(var.name, scope, var.n + scoped_name.n + 1)
         find_in_scope(new_scoped_name)
     except NameNotFound:
-        new_scoped_name = ScopedName(var.name, scope, var.n + scoped_name.n)
+        new_scoped_name = ScopedName(scoped_name.name, scope, var.n + scoped_name.n)
```
Comment (sjrl) [inline padl/dumptools/var2mod.py:626]: Why are we using `scoped_name.name` here instead of `var.name`?

File: padl/transforms.py (hunk 2 above)
Comment (jasonkhadka) [reply padl/transforms.py:529]: Then the question becomes if there are cases when the scope is not global for an object but it is in `globals_dict`. I think this can be the case when we are looking at some internal scope, where variables from the outer scope are used, but not necessarily global. 
Is this the case @wuhu? 
If not, then we can skip defining `all_vars_dict`, and use `non_local_dict` in `else` clause.

File: padl/dumptools/var2mod.py (hunk 6 above)
Comment (jasonkhadka) [reply padl/dumptools/var2mod.py]: I updated this using `var.name`.  `var.name` makes more sense.

File: padl/transforms.py (hunk 2 above)
Comment (jasonkhadka) [reply padl/transforms.py]: Thanks for the clarifaction. I reverted this change.

**Comment by jasonkhadka** [comment]
I have removed doctest on `increament_same_name_var`. The example is still there.

File: padl/dumptools/var2mod.py (hunk 1 above)
Comment (jasonkhadka) [reply padl/dumptools/var2mod.py]: i moved scope to _check_and_make_increment, and added some comment to explain the things there.

**Comment by sjrl** [comment]
@jasonkhadka What was wrong with the doctest?

**Comment by jasonkhadka** [comment]
`find_in_source` did not work with this doctest code. It is about finding the right scope. 
No fundamental issue here, just think we do not need doctest here.

> @jasonkhadka What was wrong with the doctest?

**Comment by sjrl** [comment]
@jasonkhadka Could you add the examples @wuhu showed here https://github.com/lf1-io/padl/pull/432#discussion_r832318191 as unit tests? I think it would be good to test the behavior of the `if else` statement in question.

**Review by sjrl — APPROVED** [review]
Looks good!

**Comment by jasonkhadka** [comment]
> 

yes sure!

**Comment by jasonkhadka** [comment]
> @jasonkhadka Could you add the examples @wuhu showed here https://github.com/lf1-io/padl/pull/432#discussion_r832318191 as unit tests? I think it would be good to test the behavior of the if else statement in question.

File: padl/dumptools/var2mod.py [hunk 7]
```diff
@@ -597,28 +605,52 @@ def find_globals(node: ast.AST, filter_builtins: bool = True) -> Set[Tuple[str,
...
+def _check_and_make_increment(value: ScopedName, targets: Set):
+    """Apply Increment on *value* ScopedName if necessary.
+
+    Increment means the increment of *value.n*.
+    *value.n* signifies which one of the usages of a *name* in the code does this
+     *scoped_name* points to.
+    0 being the most recent, and increase integer representing the older usages.
+
+    :param value: Target ScopedName for increment
+    :param targets: Set of ScopedName for increment
```
Comment (sjrl) [inline padl/dumptools/var2mod.py:617]: I would consider rephrasing the docstrings to something like 
```python
:param value: ScopedName to be incremented (if necessary)
:param targets: Set of ScopedName's to compare to *value*
```
Right now it's a little confusing that `value` is described as the `Target ScopedName` when we also have a parameter called `targets`.

**Review by wuhu — CHANGES_REQUESTED** [review]
This is going in the right direction, main problem I think is what I wrote to you about on slack (and here, line 621).

File: padl/dumptools/var2mod.py (hunk 7 above)
Comment (jasonkhadka) [reply padl/dumptools/var2mod.py:617]: thanks updated it.

File: padl/dumptools/var2mod.py [hunk 8]
```diff
@@ -62,13 +62,20 @@ def get_source_segments(self, source):
         ]
 
 
-def _join_attr(node):
-    if not isinstance(node, (ast.Attribute, ast.Name)):
+def _join_attr(ast_node):
+    """Get list of base object name and attribute names
+
+    'f.pd_to' -> ['f', 'pd_to']
```
[the remark this answers is not in the document]
Comment (jasonkhadka) [reply padl/dumptools/var2mod.py:68]: sure.

File: padl/dumptools/var2mod.py [hunk 9]
```diff
@@ -597,28 +605,52 @@ def find_globals(node: ast.AST, filter_builtins: bool = True) -> Set[Tuple[str,
...
+    *value.n* signifies which one of the usages of a *name* in the code does this
+     *scoped_name* points to.
+    0 being the most recent, and increase integer representing the older usages.
+
+    :param value: Target ScopedName for increment
+    :param targets: Set of ScopedName for increment
+    :return:
+        ScopedName after increment
+    """
+    split_value_name = value.name.rsplit('.', 1)[0]
```
[the remark this answers is not in the document]
Comment (jasonkhadka) [reply padl/dumptools/var2mod.py:621]: I was also thinking about this problem.
I am continuing this discussion on the issue: https://github.com/lf1-io/padl/issues/427

File: padl/dumptools/symfinder.py [hunk 10]
```diff
@@ -668,14 +681,65 @@ def replace_star_imports(tree: ast.Module):
...
+    for variant_name, variant_n in scoped_name.variants:
+        if remove_dot and '.' in variant_name:
+            variant_name = variant_name.rsplit('.', 1)[0]
+        return_scoped_name.add_variant(variant_name, add_n + variant_n)
+    return_scoped_name.variants.pop(0)
+    return_scoped_name.name, return_scoped_name.n = return_scoped_name.variants[0]
+    return return_scoped_name
+
+
+def find_scopedname_in_source(scoped_name: ScopedName, source, tree=None) -> Tuple[str, ast.AST]:
```
Comment (sjrl) [inline padl/dumptools/symfinder.py:703]: Update return type to show that `Tuple` returns three items.

File: padl/dumptools/symfinder.py [hunk 11]
```diff
@@ -668,14 +681,65 @@ def replace_star_imports(tree: ast.Module):
...
+    """Find the piece of code that assigned a value to the variable with name *var_name* in the
+    source string *source*.
+
+    :param scoped_name: ScopedName to look for.
+    :param tree: AST.Module to look into for scoped_name.
+    :param source: Source code to search.
+    :returns: Tuple with source code segment and corresponding AST node.
```
Comment (sjrl) [inline padl/dumptools/symfinder.py:710]: Update docstring to reflect returning three things.

File: padl/dumptools/symfinder.py [hunk 12]
```diff
@@ -736,6 +805,24 @@ def _find_branch(tree, lineno, source):
...
+        try:
+            source, node, name = find_scopedname_in_source(scoped_name, cell)
+        except (NameNotFound, SyntaxError):
+            continue
+        break
+    if source is None:
+        raise NameNotFound(f'{",".join([str((var, n)) for var, n in scoped_name.variants])} not found.')
+    return source, node, name
+
+
```
Comment (sjrl) [inline padl/dumptools/symfinder.py:825]: @jasonkhadka Do we need both `find_in_ipython` and `find_scopedname_in_ipython`?

File: padl/dumptools/symfinder.py [hunk 13]
```diff
@@ -607,39 +608,51 @@ def f(x):
...
+    :return: Tuple as ((source, node), scope, name), where
+        * source: String representation of piece of code.
+        * node: Ast node for the code.
+        * scope: Scope of the code.
+        * name: Name of variable (str).
+
     """
     scope = scoped_name.scope
-    i = scoped_name.n
+    n = scoped_name.n
```
Comment (sjrl) [inline padl/dumptools/symfinder.py:642]: I think you can remove this line. It looks like `n` is not used in this function.

File: padl/dumptools/symfinder.py [hunk 14]
```diff
@@ -759,6 +846,29 @@ def find_in_ipython(var_name: str, i: int = 0) -> Tuple[str, ast.AST]:
...
 
 
+def find_scopedname(scoped_name: ScopedName) -> Tuple[str, ast.AST]:
+    """Find the piece of code that assigned a value to the variable with name *var_name* in the
+    module *module*.
+
+    If *module* is not specified, this uses `__main__`. In that case, the ipython history will
+    be searched as well.
+
+    :param var_name: Name of the variable to look for.
```
Comment (sjrl) [inline padl/dumptools/symfinder.py:856]: Remove doc string for `var_name`

Comment (sjrl) [reply padl/dumptools/symfinder.py:856]: * Same for `module` and `i`.
* Add doc string for `scoped_name`

File: padl/dumptools/symfinder.py [hunk 15]
```diff
@@ -759,6 +846,29 @@ def find_in_ipython(var_name: str, i: int = 0) -> Tuple[str, ast.AST]:
     return source, node, name
 
 
+def find_scopedname(scoped_name: ScopedName) -> Tuple[str, ast.AST]:
+    """Find the piece of code that assigned a value to the variable with name *var_name* in the
+    module *module*.
```
Comment (sjrl) [inline padl/dumptools/symfinder.py:851]: Update doc string to reflect the different input variable

File: padl/dumptools/symfinder.py (hunk 13 above)
Comment (jasonkhadka) [reply padl/dumptools/symfinder.py:642]: removed!

File: padl/dumptools/symfinder.py (hunk 12 above)
Comment (jasonkhadka) [reply padl/dumptools/symfinder.py:825]: I am not sure yet, 
I don't see another use at the moment for `find_in_ipython`, but it is useful function so could be used or used for debugging.

Comment (sjrl) [reply padl/dumptools/symfinder.py:824]: I guess in general it looks like there are a few functions that look almost duplicated like `find_scopedname_in_module` and `find_in_module`, and `find_scopedname_in_source` and `find_in_source`. Is it possible to just use the new `find_scopedname_` versions?

Comment (jasonkhadka) [reply padl/dumptools/symfinder.py:824]: Yes, I updated the finders to use `find_scopedname*` instead of `find_*`, cause it contains more information. I am not sure about throwing out `find_*`s yet. They can be used in debugging at the least and perhaps will be more useful for other new features. Also, it is not a whole lot of code.

File: padl/dumptools/var2mod.py [hunk 16]
```diff
@@ -597,31 +607,74 @@ def find_globals(node: ast.AST, filter_builtins: bool = True) -> Set[Tuple[str,
...
+     *scoped_name* points to.
+    0 being the most recent, and increase integer representing the older usages.
+
+    :param value: ScopedName to be incremented (if necessary).
+    :param targets: Set of ScopedName's to compare to *value*.
+    :return: ScopedName after increment.
+    """
+    split_value_name = value.name.rsplit('.', 1)[0]
+    for target in targets:
+        split_target_name = target.name.rsplit('.', 1)[0]
```
Comment (wuhu) [inline padl/dumptools/var2mod.py:624]: I think this should cover all `.`s, not just the rightmost one.

File: padl/dumptools/var2mod.py [hunk 17]
```diff
@@ -597,31 +607,74 @@ def find_globals(node: ast.AST, filter_builtins: bool = True) -> Set[Tuple[str,
...
-    >>> isinstance(out, set)
+    >>> import sys
+    >>> a = b = 2
+    >>> b = a + b
+    >>> out = increment_same_name_var({ScopedName('a', None, 0), ScopedName('b', None, 0)},
+    ...                 ScopedName('b', Scope.toplevel(sys.modules[__name__]), 0))  # doctest: +SKIP
+    >>> isinstance(out, set)  # doctest: +SKIP
     True
-    >>> {(x.name, x.n) for x in out} == {('a', 1), ('b', 4)}
+    >>> {(x.name, x.n) for x in out} == {('a', 0), ('b', 1)}  # doctest: +SKIP
```
Comment (wuhu) [inline padl/dumptools/var2mod.py:664]: Why still skipping, does it not work?

File: padl/dumptools/symfinder.py [hunk 18]
```diff
@@ -668,14 +680,65 @@ def replace_star_imports(tree: ast.Module):
                 node.names = [ast.alias(name=name, asname=None) for name in names]
 
 
+def update_scopedname(scoped_name: ScopedName, scope: Scope, add_n: int = 0, remove_dot: bool = False):
```
Comment (wuhu) [inline padl/dumptools/symfinder.py]: You only ever use this with `add_n != 0 and remove_dot == False` or `remove_dot == True and add_n == 0`.
It would be more readable if you had two separate functions `remove_dot` and `add_n` (or something like that). (If `remove_dot` is even needed, see comment in `find_scopedname_in_source`).

File: padl/dumptools/symfinder.py [hunk 19]
```diff
@@ -607,39 +608,50 @@ def f(x):
...
+    def add_variant(self, name, n):
+        self.variants.append((name, n))
+
     def __hash__(self):
         return hash((self.name, self.scope, self.n))
 
     def __eq__(self, other):
         return (self.name, self.scope, self.n) == (other.name, other.scope, other.n)
 
+    def copy(self):
```
Comment (wuhu) [inline padl/dumptools/symfinder.py:623]: Is this ever used?

File: padl/dumptools/symfinder.py [hunk 20]
```diff
@@ -607,39 +608,50 @@ def f(x):
     scope: Scope
     n: int = 0
```
[the remark this answers is not in the document]
Comment (wuhu) [reply padl/dumptools/symfinder.py:609]: Also please add some explanation about the `variants` in the docstring.

File: padl/dumptools/var2mod.py [hunk 21]
```diff
@@ -588,6 +597,7 @@ def find_globals(node: ast.AST, filter_builtins: bool = True) -> Set[Tuple[str,
 
     :param node: AST node to search in.
     :param filter_builtins: If *True*, filter out builtins.
+    :return: Set of global ScopedNames
```
Comment (wuhu) [inline padl/dumptools/var2mod.py:600]: .

File: padl/wrap.py [hunk 22]
```diff
@@ -19,6 +19,7 @@
 )
 
 
+
```
Comment (wuhu) [inline padl/wrap.py:22]: Remove this extra line.

**Review by wuhu — CHANGES_REQUESTED** [review]

File: padl/dumptools/var2mod.py (hunk 17 above)
Comment (jasonkhadka) [reply padl/dumptools/var2mod.py:664]: yes, still does not work. Something to do with Scope.

File: padl/dumptools/symfinder.py (hunk 19 above)
Comment (jasonkhadka) [reply padl/dumptools/symfinder.py:693]: it is used with refactor.

File: padl/dumptools/symfinder.py (hunk 18 above)
Comment (jasonkhadka) [reply padl/dumptools/symfinder.py]: I have moved this to the methods of ScopedName.

File: padl/dumptools/symfinder.py [hunk 23]
```diff
@@ -668,14 +680,65 @@ def replace_star_imports(tree: ast.Module):
...
+    :param add_n: *n* to add to new ScopedName.
+    :param remove_dot: Boolean to remove dot from ScopedName or not.
+    :return: New ScopedName with updated name and *n*.
+    """
+    return_scoped_name = ScopedName(None, scope, None)
+    for variant_name, variant_n in scoped_name.variants:
+        if remove_dot and '.' in variant_name:
+            variant_name = variant_name.rsplit('.', 1)[0]
+        return_scoped_name.add_variant(variant_name, add_n + variant_n)
+    return_scoped_name.variants.pop(0)
```
[the remark this answers is not in the document]
Comment (jasonkhadka) [reply padl/dumptools/symfinder.py]: i refactored ScopedName to make it clearer.

File: padl/dumptools/symfinder.py [hunk 24]
```diff
@@ -668,14 +680,65 @@ def replace_star_imports(tree: ast.Module):
...
+    for statement in tree.body[::-1]:
+        for var_name, var_n in scoped_name.variants:
+            for finder_cls in finder_clss:
+                finder = finder_cls(source, var_name)
+                finder.visit(statement)
+                if finder.found_something():
+                    if n_dict[var_name] == 0:
+                        return finder.deparse(), finder.node(), var_name
+                    n_dict[var_name] -= 1
+    if any(['.' in var_name for var_name, _ in scoped_name.variants]):
```
[the remark this answers is not in the document]
Comment (jasonkhadka) [reply padl/dumptools/symfinder.py]: Refactored to fix this. 
Now, dump is 
```
from padl import transform


class X:
    def __init__(self, x):
        self.random = x


numpy = X(2)
_pd_main = transform(lambda x: numpy.random)
```

File: padl/dumptools/symfinder.py (hunk 12 above)
Comment (jasonkhadka) [reply padl/dumptools/symfinder.py:856]: I refactored this.

File: padl/dumptools/symfinder.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (jasonkhadka) [reply padl/dumptools/symfinder.py:618]: Right now I have kept dataclass for the convenience of the printing. I don't see any harm keeping it or big advantage removing it.

[the remark this answers is not in the document]
Comment (wuhu) [reply padl/dumptools/symfinder.py:618]: Yeah, the point was that now that the class has so much extra stuff there's also not that much of a benefit any more, and in my view if there's not much of an advantage it's better to just have an ordinary class (because people are more familiar with it).

**Comment by wuhu** [comment]
@jasonkhadka did you the "5 hidden conversations"?

**Comment by jasonkhadka** [comment]
@wuhu hadn't seen them, but the refactor takes care of all the points.

> @jasonkhadka did you the "5 hidden conversations"?

File: padl/transforms.py [hunk 25]
```diff
@@ -1692,19 +1710,21 @@ def _pd_build_codegraph(self, graph=None, name=None):
...
+
+        todo = self._pd_codegraph_add_startnodes(graph, name)
 
         # iterate over sub-transforms and update the codegraph with their codegraphs
         for transform in self.transforms:
             varname = transform.pd_varname(self._pd_call_info.scope)
             # pylint: disable=protected-access
             transform._pd_build_codegraph(graph, varname)
+
+        #self._pd_codegraph_find_dependencies(graph, todo)
```
Comment (sjrl) [inline padl/transforms.py:1727]: Consider removing commented out code
